src/Fields/fieldvector.jl

```diff
@@ -1,22 +1,92 @@
 import BlockArrays
 
+
+"""
+    FieldVector
+
+A `FieldVector` is a wrapper around one or more `Field`s that acts like vector
+of the underlying arrays.
+
+It is similar in spirit to [`ArrayPartition` from
+RecursiveArrayTools.jl](https://github.com/SciML/RecursiveArrayTools.jl#arraypartition),
+but allows referring to fields by name.
+
+# Constructors
+
+    FieldVector(;name1=field1, name2=field2, ...)
+
+Construct a `FieldVector`, wrapping `field1, field2, ...` using the names
+`name1, name2, ...`.
+"""
 struct FieldVector{T, M} <: BlockArrays.AbstractBlockVector{T}
     values::M
 end
 FieldVector{T}(values::M) where {T, M} = FieldVector{T, M}(values)
 
+
+"""
+    ScalarWrapper(val) <: AbstractArray{T,0}
+
+This is a wrapper around scalar values that allows them to be mutated as part of
+a FieldVector. A call `getproperty` on a `FieldVector` with this component will
+return a scalar, instead of the boxed object.
+"""
+mutable struct ScalarWrapper{T} <: AbstractArray{T, 0}
+    val::T
+end
+Base.size(::ScalarWrapper) = ()
+Base.getindex(s::ScalarWrapper, args...) = s.val
+Base.setindex!(s::ScalarWrapper, value, args...) = s.val = value
+Base.similar(s::ScalarWrapper) = ScalarWrapper(s.val)
+
+wrap(x) = x
+wrap(x::Real) = ScalarWrapper(x)
+unwrap(x) = x
+unwrap(x::ScalarWrapper) = x[]
+
+function FieldVector(; kwargs...)
+    values = map(wrap, NamedTuple(kwargs))
+    T = promote_type(
+        map(RecursiveArrayTools.recursive_bottom_eltype, values)...,
+    )
+    return FieldVector{T}(values)
+end
+
 _values(fv::FieldVector) = getfield(fv, :values)
-_parent_values(fv::FieldVector) = map(parent, _values(fv))
+
+"""
+    backing_array(x)
+
+The `AbstractArray` or `Ref` that is behind an object `x`, allowing it to be
+treated as a component of a `FieldVector`.
+"""
+backing_array(x) = x
```

Comment:
should we specialize this to just AbstractArray` and `Ref`?

Reply:
Oh, I meant to get rid of the `Ref` mention (since that doesn't really work for the reasons I noted above).

The input can be anything, the output should be an `AbstractArray`.